### ISSUE TITLE
Distributed training: Speed extremely slow

### ISSUE THREAD

**Issue by zzmig** [dmlc/mxnet#5158]
## Environment info
Operating System: Centos 7.0
Package used (Python/R/Scala/Julia): Python
MXNet version: 0.93
Python version and distribution: 2.7

We are testing the distributed train over multi-machine mxnet.  We compile Mxnet with parameter "USE_DIST_KVSTORE=1" , and successful run train a MLP on mnist.   But we found that the speed of distributed training was unusually slow compared to the speed of stand-alone training.

Stand-alone commands and speed：
`python train_mnist.py`
INFO:root:start with arguments Namespace(batch_size=64, disp_batches=100, gpus=None, kv_store='device', load_epoch=None, lr=0.05, lr_factor=0.1, lr_step_epochs='10', model_prefix=None, mom=0.9, monitor=0, network='mlp', num_classes=10, num_epochs=20, num_examples=60000, num_layers=None, optimizer='sgd', test_io=0, top_k=0, wd=0.0001)
INFO:root:Epoch[0] Batch [100]  Speed: 21875.54 samples/sec     Train-accuracy=0.771040
INFO:root:Epoch[0] Batch [200]  Speed: 21260.40 samples/sec     Train-accuracy=0.913906
INFO:root:Epoch[0] Batch [300]  Speed: 21302.58 samples/sec     Train-accuracy=0.933750
INFO:root:Epoch[0] Batch [400]  Speed: 21216.02 samples/sec     Train-accuracy=0.936875
INFO:root:Epoch[0] Batch [500]  Speed: 22835.21 samples/sec     Train-accuracy=0.933594
INFO:root:Epoch[0] Batch [600]  Speed: 21612.71 samples/sec     Train-accuracy=0.947500
INFO:root:Epoch[0] Batch [700]  Speed: 23362.35 samples/sec     Train-accuracy=0.954375
INFO:root:Epoch[0] Batch [800]  Speed: 21683.75 samples/sec     Train-accuracy=0.954219
INFO:root:Epoch[0] Batch [900]  Speed: 21656.14 samples/sec     Train-accuracy=0.955781
INFO:root:Epoch[0] Train-accuracy=0.958615
INFO:root:Epoch[0] Time cost=2.804
INFO:root:Epoch[0] Validation-accuracy=0.957803

Command and speed of distributed training on local machine by using two workers:
`launch.py -n 2 --launcher local python train_mnist.py --kv-store dist_sync`
INFO:root:start with arguments Namespace(batch_size=64, disp_batches=100, gpus=None, kv_store='dist_sync', load_epoch=None, lr=0.05, lr_factor=0.1, lr_step_epochs='10', model_prefix=None, mom=0.9, monitor=0, network='mlp', num_classes=10, num_epochs=20, num_examples=60000, num_layers=None, optimizer='sgd', test_io=0, top_k=0, wd=0.0001)
INFO:root:start with arguments Namespace(batch_size=64, disp_batches=100, gpus=None, kv_store='dist_sync', load_epoch=None, lr=0.05, lr_factor=0.1, lr_step_epochs='10', model_prefix=None, mom=0.9, monitor=0, network='mlp', num_classes=10, num_epochs=20, num_examples=60000, num_layers=None, optimizer='sgd', test_io=0, top_k=0, wd=0.0001)
INFO:root:Epoch[0] Batch [100]  Speed: 2580.90 samples/sec      Train-accuracy=0.104425
INFO:root:Epoch[0] Batch [100]  Speed: 2574.24 samples/sec      Train-accuracy=0.096689
INFO:root:Epoch[0] Batch [200]  Speed: 4408.62 samples/sec      Train-accuracy=0.093594
INFO:root:Epoch[0] Batch [200]  Speed: 4399.50 samples/sec      Train-accuracy=0.099531
INFO:root:Epoch[0] Batch [300]  Speed: 3343.81 samples/sec      Train-accuracy=0.095469
INFO:root:Epoch[0] Batch [300]  Speed: 3342.38 samples/sec      Train-accuracy=0.097344
INFO:root:Epoch[0] Batch [400]  Speed: 3247.69 samples/sec      Train-accuracy=0.096250
INFO:root:Epoch[0] Batch [400]  Speed: 3245.97 samples/sec      Train-accuracy=0.104531
INFO:root:Epoch[0] Batch [500]  Speed: 3364.46 samples/sec      Train-accuracy=0.102188
INFO:root:Epoch[0] Batch [500]  Speed: 3364.97 samples/sec      Train-accuracy=0.098437
INFO:root:Epoch[0] Batch [600]  Speed: 3729.89 samples/sec      Train-accuracy=0.097500
INFO:root:Epoch[0] Batch [600]  Speed: 3732.76 samples/sec      Train-accuracy=0.097812
INFO:root:Epoch[0] Batch [700]  Speed: 5105.08 samples/sec      Train-accuracy=0.087969
INFO:root:Epoch[0] Batch [700]  Speed: 5097.00 samples/sec      Train-accuracy=0.104375
INFO:root:Epoch[0] Batch [800]  Speed: 3931.05 samples/sec      Train-accuracy=0.099062
INFO:root:Epoch[0] Batch [800]  Speed: 3930.63 samples/sec      Train-accuracy=0.101406
INFO:root:Epoch[0] Batch [900]  Speed: 3763.65 samples/sec      Train-accuracy=0.098906
INFO:root:Epoch[0] Batch [900]  Speed: 3758.21 samples/sec      Train-accuracy=0.099531
INFO:root:Epoch[0] Train-accuracy=0.104307
INFO:root:Epoch[0] Time cost=16.560
INFO:root:Epoch[0] Train-accuracy=0.099662
INFO:root:Epoch[0] Time cost=16.584
INFO:root:Epoch[0] Validation-accuracy=0.098029
INFO:root:Epoch[0] Validation-accuracy=0.098029

Command and speed of distributed training on two computers(with same hardware configration):
`launch.py -n 2 --launcher ssh -H hosts python train_mnist.py --kv-store dist_sync`
INFO:root:start with arguments Namespace(batch_size=64, disp_batches=100, gpus=None, kv_store='dist_sync', load_epoch=None, lr=0.05, lr_factor=0.1, lr_step_epochs='10', model_prefix=None, mom=0.9, monitor=0, network='mlp', num_classes=10, num_epochs=20, num_examples=60000, num_layers=None, optimizer='sgd', test_io=0, top_k=0, wd=0.0001)
INFO:root:start with arguments Namespace(batch_size=64, disp_batches=100, gpus=None, kv_store='dist_sync', load_epoch=None, lr=0.05, lr_factor=0.1, lr_step_epochs='10', model_prefix=None, mom=0.9, monitor=0, network='mlp', num_classes=10, num_epochs=20, num_examples=60000, num_layers=None, optimizer='sgd', test_io=0, top_k=0, wd=0.0001)
INFO:root:Epoch[0] Batch [100]  Speed: 614.03 samples/sec       Train-accuracy=0.097463
INFO:root:Epoch[0] Batch [100]  Speed: 613.66 samples/sec       Train-accuracy=0.096225
INFO:root:Epoch[0] Batch [200]  Speed: 577.24 samples/sec       Train-accuracy=0.097812
INFO:root:Epoch[0] Batch [200]  Speed: 576.24 samples/sec       Train-accuracy=0.096250
INFO:root:Epoch[0] Batch [300]  Speed: 609.78 samples/sec       Train-accuracy=0.094531
INFO:root:Epoch[0] Batch [300]  Speed: 610.73 samples/sec       Train-accuracy=0.098125
INFO:root:Epoch[0] Batch [400]  Speed: 624.24 samples/sec       Train-accuracy=0.100000
INFO:root:Epoch[0] Batch [400]  Speed: 623.74 samples/sec       Train-accuracy=0.100625
INFO:root:Epoch[0] Batch [500]  Speed: 642.81 samples/sec       Train-accuracy=0.096719
INFO:root:Epoch[0] Batch [500]  Speed: 643.28 samples/sec       Train-accuracy=0.099531
INFO:root:Epoch[0] Batch [600]  Speed: 613.65 samples/sec       Train-accuracy=0.097187
INFO:root:Epoch[0] Batch [600]  Speed: 613.63 samples/sec       Train-accuracy=0.096406
INFO:root:Epoch[0] Batch [700]  Speed: 626.13 samples/sec       Train-accuracy=0.106875
INFO:root:Epoch[0] Batch [700]  Speed: 626.15 samples/sec       Train-accuracy=0.096562
INFO:root:Epoch[0] Batch [800]  Speed: 621.21 samples/sec       Train-accuracy=0.098437
INFO:root:Epoch[0] Batch [800]  Speed: 620.83 samples/sec       Train-accuracy=0.099687
INFO:root:Epoch[0] Batch [900]  Speed: 589.75 samples/sec       Train-accuracy=0.098281
INFO:root:Epoch[0] Batch [900]  Speed: 589.72 samples/sec       Train-accuracy=0.104844
INFO:root:Epoch[0] Train-accuracy=0.101351
INFO:root:Epoch[0] Time cost=98.129
INFO:root:Epoch[0] Train-accuracy=0.097551
INFO:root:Epoch[0] Time cost=98.144
INFO:root:Epoch[0] Validation-accuracy=0.098029
INFO:root:Epoch[0] Validation-accuracy=0.098029
INFO:root:Epoch[1] Batch [100]  Speed: 639.28 samples/sec       Train-accuracy=0.097308
INFO:root:Epoch[1] Batch [100]  Speed: 640.32 samples/sec       Train-accuracy=0.096844

We do not know why the speed so slow in distributed training , someone can help us? 
Thanks

**Comment by mli** [comment]
please try a larger workload, such as `train_cifar10`. and also there are some suggestions: https://github.com/mli/mxnet/tree/master/example/image-classification#speed

**Comment by piiswrong** [comment]
@mli Let's add a warning somewhere to tell people training small networks on multiple machines doesn't make sense.

**Comment by mli** [comment]
good idea.

**Comment by wlike** [comment]
I am experimenting with MXNMT from https://github.com/magic282/MXNMT.
The training data is 1million. I have 2 machines each with 8 GPUs (I only use the first 6).

When I launch the job with `/path/to/launch.py -n 2 --launcher ssh -H hosts python main.py`, and `batch_size=360(each gpu 60)`, `kvstore=dist_sync`, the performance is:
![image](https://cloud.githubusercontent.com/assets/6084282/23505194/7cd63724-ff7d-11e6-81ea-284776b142c3.png)

When with `/path/to/launch.py -n 2 -s 4 --launcher ssh -H hosts python main.py`, the performance is:
![image](https://cloud.githubusercontent.com/assets/6084282/23505240/b004e622-ff7d-11e6-8b22-2b53f5e1d015.png)

When with `/path/to/launch.py -n 2 -s 6 --launcher ssh -H hosts python main.py`, the performance is:
![image](https://cloud.githubusercontent.com/assets/6084282/23505277/dbc8b950-ff7d-11e6-842d-3fb83d36ba23.png)

My question is how to set server number, and among these servers, which are level 1 servers, which are level 2 servers (According to https://github.com/dmlc/mxnet/issues/797).
